### PR TITLE
[SPARK-23608][CORE][WebUI] Add synchronization in SHS between attachSparkUI and detachSparkUI functions to avoid concurrent modification issue to Jetty Handlers

### PR DESCRIPTION
**Pull request by zhouyejoe**
Jetty handlers are dynamically attached/detached while SHS is running. But the attach and detach operations might be taking place at the same time due to the async in load/clear in Guava Cache. 

## What changes were proposed in this pull request?
Add synchronization between attachSparkUI and detachSparkUI in SHS.

## How was this patch tested?
With this patch, the jetty handlers missing issue never happens again in our production cluster SHS.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 1]
```diff
@@ -148,14 +148,14 @@ class HistoryServer(
...
-      completed: Boolean) {
+      completed: Boolean) = this.synchronized {
     assert(serverInfo.isDefined, "HistoryServer must be bound before attaching SparkUIs")
     ui.getHandlers.foreach(attachHandler)
     addFilters(ui.getHandlers, conf)
   }
 
   /** Detach a reconstructed UI from this server. Only valid after bind(). */
-  override def detachSparkUI(appId: String, attemptId: Option[String], ui: SparkUI): Unit = {
+  override def detachSparkUI(appId: String, attemptId: Option[String], ui: SparkUI): Unit = this.synchronized {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: I'll trigger tests but this will cause a style violation.

**Comment by vanzin** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala (hunk 1 above)
Comment (zhouyejoe) [reply core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Thanks. Will fix the style violation.

**Comment by zhouyejoe** [comment]
Updated. Please help trigger the jenkins again. Thanks.

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 2]
```diff
@@ -148,14 +148,17 @@ class HistoryServer(
       appId: String,
       attemptId: Option[String],
       ui: SparkUI,
-      completed: Boolean) {
+      completed: Boolean): Unit = this.synchronized {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Hmm... would it be better to synchronize `SparkUI.attachHandler` and `SparkUI.detachHandler`? Or maybe restrict the scope of the synchronization here?

This seems too coarse; in the detach case, it synchronizes around the `onUIDetached` callback which can do expensive locking and I/O.

Comment (zhouyejoe) [reply core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: In order to fine grain the synchronization, can we add synchronize on handlers only?

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 3]
```diff
@@ -150,14 +150,18 @@ class HistoryServer(
       ui: SparkUI,
       completed: Boolean) {
     assert(serverInfo.isDefined, "HistoryServer must be bound before attaching SparkUIs")
-    ui.getHandlers.foreach(attachHandler)
+    handlers.synchronized {
+      ui.getHandlers.foreach(attachHandler)
+    }
     addFilters(ui.getHandlers, conf)
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: shall we also keep `handlers` synchronized for this line?

Comment (zhouyejoe) [reply core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Yes. Thanks for catching this. Will update.

**Comment by zhouyejoe** [comment]
It is weird that the PySpark unit tests failed, I don't think it is related.

**Comment by vanzin** [comment]
This doesn't have anything to do with pyspark, so we can ignore those.

Merging to master / 2.3.

**Pull request closed without merging**